src/init.cpp

```diff
@@ -229,6 +229,9 @@ void Shutdown()
         if (pcoinsTip != NULL) {
             FlushStateToDisk();
         }
+        // Flush the wallet witness cache to disk (no longer done by FlushStateToDisk())
+        GetMainSignals().SetBestChain(chainActive.GetLocator());
```

Comment:
Before this PR, the witnesses were not being flushed to disk even on a clean shutdown, which meant that every time you restart, the blocks that were added since the most recent one-hour flush were rescanned (`init message: Rescanning...` in debug.log). With this change, following a clean shutdown no rescanning occurs. Rescanning an hour's worth of blocks is pretty fast, but this is still cleaner.

Reply:
Nit, not related to this PR: the fact that SetBestChain performs a flush seems like a bit of a smell. I really like IO to be explicit.